orttraining/orttraining/python/training/ortmodule/_utils.py

```diff
@@ -38,18 +39,46 @@ def _ortvalue_from_torch_tensor(torch_tensor):
     return C.OrtValue.from_dlpack(to_dlpack(torch_tensor), is_bool_tensor)
 
 
-def _torch_tensor_from_dl_pack(dlpack, ortvalue, device):
-    torch_tensor = from_dlpack(dlpack) if device.type != 'ort' else C.ort_from_dlpack(dlpack)
-    return torch_tensor.to(torch.bool) if ortvalue.data_type() == 'tensor(bool)' else torch_tensor
+def _ortvalues_to_torch_tensor(ortvalues, device):
+    if len(ortvalues) == 0:
+        return tuple()
+    if hasattr(ortvalues, 'to_dlpack'):
+        if all(ortvalue.proto_type() != 9 for ortvalue in ortvalues):
```

Comment:
If we agree that the new 'to_dlpack' works well and fast, then even ortvalue.proto_type() is 9, we can still use the new 'to_dlpack' to get all torch.Tensors, then convert those torch tensors at position that their corresponding OrtValues are bool to bool torch tensors. So we don't need the `if hasattr(ortvalues, 'dlpack_at'):` clause below anymore. Also suggest use Enum name instead of number (9) for better code reading.